frontend/src/app/data-entry/location-form/location-source-text-form/location-source-text-form.component.ts

```diff
@@ -0,0 +1,135 @@
+import { Component, DestroyRef, OnDestroy, OnInit } from "@angular/core";
+import { takeUntilDestroyed } from "@angular/core/rxjs-interop";
+import { FormGroup, FormControl } from "@angular/forms";
+import { ToastService } from "@services/toast.service";
+import { MutationResult } from "apollo-angular";
+import {
+    DataEntryLocationSourceTextGQL,
+    DataEntryUpdateLocationGQL,
+    DataEntryUpdateLocationMutation,
+} from "generated/graphql";
+import {
+    switchMap,
+    map,
+    shareReplay,
+    BehaviorSubject,
+    filter,
+    debounceTime,
+    share,
+    withLatestFrom,
+} from "rxjs";
+import { FormService } from "../../shared/form.service";
+import { FormStatus } from "../../shared/types";
+
+@Component({
+    selector: "lc-location-source-text-form",
+    templateUrl: "./location-source-text-form.component.html",
+    styleUrls: ["./location-source-text-form.component.scss"],
+})
+export class LocationSourceTextFormComponent implements OnInit, OnDestroy {
+    private id$ = this.formService.id$;
+
+    public location$ = this.id$.pipe(
+        switchMap((id) => this.spaceQuery.watch({ id }).valueChanges),
+        map((result) => result.data.spaceDescription),
+        shareReplay(1)
+    );
+
+    public form = new FormGroup({
+        designators: new FormControl<string[]>([], {
+            nonNullable: true,
+        }),
+        book: new FormControl<string>("", {
+            nonNullable: true,
+        }),
+        chapter: new FormControl<string>("", {
+            nonNullable: true,
+        }),
+        page: new FormControl<string>("", {
+            nonNullable: true,
+        }),
+    });
+
+    private formName = "sourceText";
+    private status$ = new BehaviorSubject<FormStatus>("idle");
+
+    constructor(
+        private destroyRef: DestroyRef,
+        private formService: FormService,
+        private toastService: ToastService,
+        private spaceQuery: DataEntryLocationSourceTextGQL,
+        private updateLocation: DataEntryUpdateLocationGQL
+    ) {}
+
+    ngOnInit(): void {
+        this.formService.attachForm(this.formName, this.status$);
+
+        this.location$
+            .pipe(takeUntilDestroyed(this.destroyRef))
+            .subscribe((location) => {
+                if (!location) {
+                    return;
+                }
+                this.form.patchValue(location, {
+                    emitEvent: false,
+                    onlySelf: true,
+                });
+            });
+
+        this.form.statusChanges
+            .pipe(
+                filter((status) => status == "INVALID"),
+                takeUntilDestroyed(this.destroyRef)
+            )
+            .subscribe(() => this.status$.next("invalid"));
+
+        const validFormSubmission$ = this.location$.pipe(
```

Comment:
It's not clear to my why this is piped on `this.location$` - since looks like you're immediately switching to `() => this.form.valueChanges`, what do you need the location observable for?

Reply:
I wanted to only start firing mutations on `form.valueChanges` after the initial data has been loaded in. My idea was that this way, the user can't really change anything if the internet connection is slow and the data is not fully loaded in yet. Originally, it was meant to prevent the initial load from triggering a mutation, but `emitEvent: false` and `onlySelf: true` in `location$.subscribe` also accomplish this.